### PR TITLE
chore: Add guest email blacklist

### PR DESCRIPTION
**Pull request by keithwillcode**
## What does this PR do?

Provides a way to disallow emails from being added as guest to bookings

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] (N/A) - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)
  - Reasoning: This is urgent and we need to ship ASAP. We'll come back and add tests

## How should this be tested?

- Set this env var and then try to add the email in the blacklist as a guest. It should be ignored.

### CODE REVIEW

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 1]
```diff
@@ -1398,7 +1398,16 @@ async function handler(
...
   ];
 
+  const blacklistedGuestEmails = process.env.BLACKLISTED_GUEST_EMAILS
+    ? process.env.BLACKLISTED_GUEST_EMAILS.split(",")
+    : [];
+
+  const guestsRemoved: string[] = [];
   const guests = (reqGuests || []).reduce((guestArray, guest) => {
+    if (blacklistedGuestEmails.some((e) => e === guest)) {
+      guestsRemoved.push(guest);
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking.ts:1408]: We should handle variations of the same email address e.g `email+something` and specifically for Gmail we might need to think about ignoring dots.(Detectable as @gmail.com but work emails give no clue about it)

image

**Review by PeerRich — DISMISSED** [review]
i think this is a better approach: https://github.com/calcom/cal.com/pull/15253

more long-term sustainable and doenst require manually adding emails

happy to merge both but i think https://github.com/calcom/cal.com/pull/15253 is better

File: packages/features/bookings/lib/handleNewBooking.ts (hunk 1 above)
Comment (keithwillcode) [reply packages/features/bookings/lib/handleNewBooking.ts:1408]: Yeah agreed. This PR was a very quick solution just to get a stop-gap in place. I think we can build more on top as we go.

**Comment by keithwillcode** [comment]
@PeerRich Agreed. This was merely a stop-gap to have something deployed ASAP. We can iterate as needed. https://github.com/calcom/cal.com/pull/15253 is lacking a solution for guests at the moment and this PR is lacking a solution for the main email. Perhaps we ship both or we update the other one to remove guest emails that aren't verified.

**Comment by keithwillcode** [comment]
> @PeerRich Agreed. This was merely a stop-gap to have something deployed ASAP. We can iterate as needed. #15253 is lacking a solution for guests at the moment and this PR is lacking a solution for the main email. Perhaps we ship both or we update the other one to remove guest emails that aren't verified.

Edit - See comment [here](https://github.com/calcom/cal.com/pull/15253#issuecomment-2139912371) on why I think moving forward with this solution modified slightly is preferable, at least in the short-term.

**Comment by Udit-takkar** [comment]
When BLACKLISTED_GUEST_EMAILS="pro@example.com"

1) you are not logged in 

https://github.com/calcom/cal.com/assets/53316345/52fc92be-8fd2-4bea-8dcb-78aae83c7d88

2) Logged in as different user + also prevents pro+2@example.com, pro+3@example.com


https://github.com/calcom/cal.com/assets/53316345/5f12c3b2-abbf-405b-8e1c-bb5c8f508baf

File: packages/trpc/server/routers/publicViewer/checkIfUserEmailVerificationRequired.handler.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    (guestEmail: string) => guestEmail.toLowerCase() === baseEmail.toLowerCase()
+  );
+
+  if (!!blacklistedEmail && blacklistedEmail !== userSessionEmail) {
+    log.warn(`blacklistedEmail: ${blacklistedEmail}`);
+    return true;
+  }
+  return false;
+};
+
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/publicViewer/checkIfUserEmailVerificationRequired.handler.ts:32]: Matches email with BLACKLISTED_GUEST_EMAILS and if found then check the logged in user is same as blacklisted email.

if logged in user email is not same as blacklisted email then we require verification

**Review by Udit-takkar — DISMISSED** [review]

**Review by Udit-takkar — APPROVED** [review]
LGTM

**Review by zomars — APPROVED** [review]

**Pull request merged**